### DISCUSSION TITLE
Allow Attribute on anonymous type and it's members(property).

### DISCUSSION THREAD

**Discussion by dvbava** [dotnet/csharplang#5475] [Language Ideas]
I would like to see  a new feature which allow me to use Attribute on anonymous type and it's property.

This is useful and time saving feature when doing one off small task( for ex serialisation when making HTTP call to rest endpoint).

at present we are forced to full type if we required to use Attribute.
example below is best explanation.

making HTTP POST request with payload where you need to serialise your payload  in json with different property name.

```    
    public class PaymentRequest
    {
        [JsonPropertyName("checkout_reference")]
        public string CheckoutReference { get; set; }

        [JsonPropertyName("payment_amount")]
        public double Amount { get; set; }

        [JsonPropertyName("currency")]
        public string CurrencyISO { get; set; }
    }
 
    var request = new PaymentRequest { CheckoutReference ="abc", ... }

    var payload  = JsonSerializer.Serialize(request )

```
     
Now **PaymentRequest** class always stays in project in a separate "PaymentRequest.cs" file.

However if we allow property attribute on anonymous types we can shorten the entire thing.

```
    var payload  = JsonSerializer.Serialize(new {  [JsonPropertyName("checkout_reference")]checkout_reference = "abc", ... })

```

**Comment by svick** [comment]
I feel like this doesn't save that much typing, results in messy-looking code and wouldn't be useful that often. Also, you can shorten your code a bit using records:

```c#
public record PaymentRequest(
    [property: JsonPropertyName("checkout_reference")]
    string CheckoutReference,

    [property: JsonPropertyName("payment_amount")]
    double Amount,

    [property: JsonPropertyName("currency")]
    string CurrencyISO
);

var request = new PaymentRequest("abc", ...);

var payload = JsonSerializer.Serialize(request);
```

So this feature doesn't sound particularly worthwhile to me.

**Comment by dvbava** [comment]
I do not understand how one liner code would look messy! 
The whole point is not to create the full permanent type not even Record or Class as both are full type.

Your version is no different other then you have used record(instead of class) and is much more messy. it need unnecessary creating permanent full types(record in this case), then subsequently initialising them before use.

**Reply by koszeggy** [reply]
It's not quite clear to me what you mean a 'permament class' but anonymous types are just as [permanent](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDCpA3oaQ+WRiqQLIAUAlLfY/wDcAhlFJCAdhHEBPALYBJcQGcALhLAxSAXlLiYAd1qkACjm2kARMIA2AVxgWANCazmUrgL4BuPgw+EPIA==) as records. Note that the generated `<>f__AnonymousType0<<P1>j__TPar, <P2>j__TPar>` class is an internal one.

So using a private record within your parent class will actually generate a cleaner type with even narrower visibility than an anonymous type.

**Reply by dvbava** [reply]
'permament class' is part of your source code. anonymous types coming to existence permanently after compilation(transparently) and is not concern to developer as compiler generates and manages internally.. same as lock , using, foreach etc getting trans-piled too.

for end developers, anonymous types are handy, convenient, created when needed on the fly and alternatives to full types(or Records for that matter)

**Reply by svick** [reply]
> I do not understand how one liner code would look messy!

In your fairly simple example, that one line would be over 200 characters long and I think that's too much. If you format it to multiple lines, it would look something like this:

```c#
var payload  = JsonSerializer.Serialize(new {
    [JsonPropertyName("checkout_reference")] CheckoutReference = "abc",
    [JsonPropertyName("payment_amount")] Amount  = 9,
    [JsonPropertyName("currency")] CurrencyISO = "USD"
});
```

It's definitely subjective, but I think this doesn't look great.

Though it raises the question: why are you attempting to use `JsonPropertyName` here in the first place? Doesn't code like the following produce the output you want, while adhering to your constraints?

```c#
var payload = JsonSerializer.Serialize(new { checkout_reference = "abc", payment_amount = 9.99, currency = "USD" });
```

**Reply by dvbava** [reply]
It may be that I may have picked up incorrect example here, however the point remain same.
